pilot/pkg/xds/ep_filters.go

```diff
@@ -61,73 +60,91 @@ func (b *EndpointBuilder) EndpointsByNetworkFilter(endpoints []*LocLbEndpointsAn
 			},
 		}
 
-		// Weight (number of endpoints) for the EDS cluster for each remote networks
-		remoteEps := map[network.ID]uint32{}
-		// Calculate remote network endpoints
+		// Create a map to keep track of the gateways used and their aggregate weights.
+		gatewayWeights := make(map[model.NetworkGateway]uint32)
+
+		// Process all of the endpoints.
 		for i, lbEp := range ep.llbEndpoints.LbEndpoints {
+			// Copy the endpoint in order to expand the load balancing weight. This will allow
+			// us to more easily spread traffic to the endpoint across multiple network gateways, increasing
+			// reliability of the endpoint. When multiplying, be careful to avoid overflow - clipping the
+			// result at the maximum value for uint32.
+			weight := b.scaleEndpointLBWeight(lbEp, endpointLBWeightScaleFactor)
```

Comment:
Two comments:
* For the most part the changes here and the problem you're solving could be dealt with simply by filtering the gateways before calculating multiples/performing filtering _but_  it may be easier to understand (took me a while to wrap my head around the original logic)
* I'm not confident a constant scale factor is a good idea. We _should_ be able to calculate one that guarantees divisibility by len(gateways):
   * `factorial(max(possibleGatewayLens))`

Reply:
> For the most part the changes here and the problem you're solving could be dealt with simply by filtering the gateways before calculating multiples/performing filtering but it may be easier to understand (took me a while to wrap my head around the original logic)

Agreed, refactoring the logic wasn't necessary, but I think it is an improvement for understandability and future maintenance as a result.

> I'm not confident a constant scale factor is a good idea. We should be able to calculate one that guarantees divisibility by len(gateways)

Do you have a specific case in mind as to why it might not be a good idea? This is a reasonably hot data path, so I'd like to avoid a lot of extra traversals if possible. Acknowledge that 4 is a bit of a magic number, but it's also reasonably high WRT spreading traffic across gateways on the same network or network+cluster.

Reply:
The second we hit 5 gateways we can have issues where we don't maintain the original balance (if only slightly). It's the sort of subtle problem that would be hard to diagnose in the future. Systems or vendors that prefer several gateways for reliability could be quick to cause issues. 

The factorial logic isn't very different from our original multiplier calculation, if anything it's slightly quicker and would produce a smaller common factor. 

I don't feel _too_ strongly, but I don't think the cost is high. We can check with benchmarks.